Added nxos_lldp_global resource module

adharshsrivatsr opened the pull request:
##### SUMMARY
Add nxos_lldp_global resource module

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
modules/network/nxos/nxos_lldp_global.py

adharshsrivatsr committed:
Added nxos_lldp_global module

Tests

Correction

remove q

adharshsrivatsr committed:
Added tests

adharshsrivatsr committed:
Corrections

Changes

Last few changes please

Final

Finall

Finalll

adharshsrivatsr committed:
Key error and punctuation changes

adharshsrivatsr committed:
.gitignore edit

adharshsrivatsr committed:
Facts file edit

trishnaguha commented on lib/ansible/module_utils/network/nxos/config/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,244 @@
+        want = self._module.params['config']
+        have = existing_lldp_global_facts
+        if len(have) == 0:
+            have = {}
```
This isn't necessary as facts will always be empty if running-config has no lldp config.

adharshsrivatsr replied:
I have used this so that, for a fresh setup, have will return { } and not [ ].

trishnaguha commented on lib/ansible/module_utils/network/nxos/config/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,244 @@
+        if len(have) == 0:
+            have = {}
+        resp = self.set_state(remove_empties(want), have)
+        return resp.split('\n')[:-1]
```
This should be handled in the method where commands are being generated.

trishnaguha commented on lib/ansible/module_utils/network/nxos/config/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,244 @@
+                command = 'lldp ' + key + ' ' + str(val) + '\n'
+        return command
+
+    def process_nested_dict(self, val):
```
use the `flatten_dict` method from module_utils.network.nxos.utils.utils.py

trishnaguha commented on lib/ansible/module_utils/network/nxos/config/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,244 @@
+        """
+        commands = ''
+        for key, val in have.items():
+            if 'holdtime' in key:
```
set the parameters to `default` by prepending `no`, not appending default values.

trishnaguha commented on lib/ansible/module_utils/network/nxos/config/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,244 @@
+        havek = self.get_nested_keys(have, havek)
+        include_params = list(set(havek) - set(wantk))  # keys in have and not in want; they need to be defaulted
+        for param in include_params:
+            commands += 'lldp '
```
Just call deleted operation instead of repetition of code.

trishnaguha commented on lib/ansible/module_utils/network/nxos/config/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,244 @@
+        havek = have.keys()
+        wantk = self.get_nested_keys(want, wantk)
+        havek = self.get_nested_keys(have, havek)
+        include_params = list(set(havek) - set(wantk))  # keys in have and not in want; they need to be defaulted
```
Write a method to generate final diff so it can be used anywhere in the code when required.

trishnaguha commented on lib/ansible/module_utils/network/nxos/config/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,244 @@
+        :returns: the commands necessary to migrate the current configuration
+                  to the desired configuration
+        """
+        commands = ''
```
`commands` should always be a list.

trishnaguha commented on lib/ansible/module_utils/network/nxos/facts/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,112 @@
+        data = re.split('\n', conf)
+        if len(data) > 1:
+            for key in data:
+                words = key.split()
```
Use regular expression instead of string split.

trishnaguha commented on lib/ansible/modules/network/nxos/nxos_lldp_global.py:
```diff
@@ -0,0 +1,249 @@
+description: This module configures and manages the Link Layer Discovery Protocol(LLDP) attributes on NX-OS platforms.
+author: Adharsh Srivats Rangarajan (@adharshsrivatsr)
+notes:
+  - Tested against <network_os> 7.3.(0)D1(1) on VIRL
```
`<network_os>` should be NXOS

trishnaguha replied:
In the notes, Add that `feature lldp` has to be enabled in order to use this module.

trishnaguha commented on lib/ansible/modules/network/nxos/nxos_lldp_global.py:
```diff
@@ -0,0 +1,249 @@
+  description: The set of commands pushed to the remote device.
+  returned: always
+  type: list
+  sample: ['command 1', 'command 2', 'command 3']
```
Give actual lldp commands here as sample that are being pushed to the device.

trishnaguha commented on lib/ansible/module_utils/network/nxos/facts/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,112 @@
+            facts['lldp_global'] = params['config']
+            facts = utils.remove_empties(facts)
+        else:
+            facts['lldp_global'] = {}
```
Don't return lldp_global if no facts is available.

trishnaguha reviewed (changes requested):
Needs rebase. Please update the module code as per the review comments.

adharshsrivatsr committed:
Changed facts collection and 'replaced' state

Changed facts collection and replaced state

Changed facts collection and replaced state

adharshsrivatsr committed:
Spacing and module error correction

trishnaguha commented on lib/ansible/module_utils/network/nxos/config/lldp_global/lldp_global.py:
```diff
@@ -0,0 +1,244 @@
+        if not have:
+            have = {}
+        if not want:
+            want = {}
```
No need to set `have`, `want` explicitly as it is always be empty dict if not resource is given/present.

trishnaguha committed:
return facts as empty dict not list

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>
